### PR TITLE
[JENKINS-68568] Require Java 11 or newer

### PR DESCRIPTION
**Pull request by basil**
[Developer mailing list thread](https://groups.google.com/g/jenkinsci-dev/c/pjfvsMw-EMM/m/OIw_hveUBwAJ)

### CODE REVIEW

File: pom.xml
Comment (jglick) [inline pom.xml:907]: No this must be reverted so long as https://github.com/jenkinsci/plugin-pom/blob/3325bb8e8a67c192b97fc6fe130265736d6480e1/pom.xml#L64

Comment (basil) [reply pom.xml:907]: It cannot be reverted while also accomplishing the purpose of this PR, which is stated in the PR description:

> Drop support for Java 8

The minimum Jenkins version cannot yet be updated because we do not know which version will drop support for Java 8, but the intention is that when we do we would update the minimum Jenkins version to that version.

File: src/it/benchmark/pom.xml
Comment (basil) [reply src/it/benchmark/pom.xml:17]: It cannot be reverted while also accomplishing the purpose of this PR, which is stated in the PR description:

> Drop support for Java 8

The minimum Jenkins version cannot yet be updated because we do not know which version will drop support for Java 8, but the intention is that when we do we would update the minimum Jenkins version to that version.

File: pom.xml
Comment (basil) [reply pom.xml:493]: Good thing this PR explicitly drops support for `java.level=8`, then, as stated in the PR description:

> Drop support for Java 8

Comment (jglick) [reply pom.xml:493]: Yeah but we cannot just do that because then people cannot simply accept updates to the parent POM without also updating `jenkins.version` to something very new.

**Comment by basil** [comment]
> `java.level` seems problematic, though. You may not define it to 11 when using a pre-switch `jenkins.version`.

Right, which is why this PR drops support for `java.level=8`.

File: pom.xml
Comment (jglick) [reply pom.xml:493]: I mean, we could, but then we have plugin POM versions that work with older cores but are not well suited to newer cores, and POM versions that require newer cores, without any overlap. So this seems pretty drastic when we normally encourage people to update the parent even when the core version is a few LTS lines behind. For example, anyone staying with an older line would not be able to accept new POM versions for totally unrelated reasons (new tools or JTH).

If we really go with the flag day approach in this PR, then we should definitely get rid of the `java.level` property as it serves no purpose. Add a profile making it an error to set it, and hard-code 11 as the Java level. And make sure that the build fails early and clearly when your `jenkins.version` is too old.

Comment (basil) [reply pom.xml:493]: I think we can. We can always do a backport branch if we want to deliver plugin parent POM updates to plugins that do not want to upgrade `jenkins.version` to something very new.

Comment (basil) [reply pom.xml:493]: > So this seems pretty drastic

Yes it is drastic but nevertheless feasible without changing the current semantics of `java.level` being defined in a `<properties>` section in plugin `pom.xml` files. Other proposals like #133 are not feasible as far as I can tell so long as feedback like [this](https://github.com/jenkinsci/plugin-pom/pull/133#discussion_r359344017) remain unaddressed.

> Instead figure out what Java level to pass to tools based on the jenkins.version (somehow).

That approach could be feasible (and less drastic!), but it would be a lot more work and (as of today) I am explicitly not volunteering to do that work.

**Comment by basil** [comment]
To summarize the status at this time, there is:

- My proposal in this PR, which is feasible and implemented, but involves a drastic flag day (not ideal)
- @jglick's proposal to drop `java.level` from plugin `pom.xml` files in favor of inferring this based on `jenkins.version`, which seems feasible and does not involve a drastic flag day (seems better!), but which nobody has either implemented or volunteered to implement
- @oleg-nenashev's proposal to use a conditional profile activated with `java.level`, which as far as I can tell is not feasible so long as feedback like [this](https://github.com/jenkinsci/plugin-pom/pull/133#discussion_r359344017) remains unaddressed

> Add […], and hard-code […]. And make sure that […].

Based on the use of the imperative mood in the above comments, I would like to clarify my intention when filing this PR. My intention was to do the minimum amount of work to reach a feasible result, which I believe I have accomplished. I am not against anyone doing a more complete and ideal implementation in another PR; however, my goal for this PR was merely to reach a minimum viable product (MVP).

File: pom.xml
Comment (timja) [reply pom.xml:493]: Agree with Basil, because of how this was done you can't keep compat.

We can run a 4.x branch if we need to backport changes...

alternatively plugins that want to stay on old core could just set this to 1.8?

**Review by timja — APPROVED** [review]
LGTM once `jenkins.version` is updated (when we figure out what that is)

File: pom.xml
Comment (jglick) [reply pom.xml:493]: > set this to 1.8

IIRC that will break _other_ places which expect `8`.

**Comment by jglick** [comment]
Since it seems like the easiest option to implement, and cleans up tech debt, I am OK with a “flag day” switch like this, so that people updating to a post-switch `jenkins.version` could (must?) update the parent, while people sticking to a pre-switch `jenkins.version` must also stick with an old parent. (Or perhaps some backport versions, if we identify sufficiently important fixes we want everyone to adopt.) We do however need to ensure that if you inadvertently update the parent (e.g., Dependabot) while still using a pre-switch `jenkins.version` (and presumably `<java.level>8</java.level>`) that the build fails and that the error message is comprehensible, so devs will be forced to look at the release notes and decide what to do.

I do recommend that as part of this we drop the `java.level` property, which was never actually overridable in a meaningful way—you could make your plugin require Java 11 when core is 8-compatible but the Jenkins plugin manager running on 8 would AFAIK not protect users properly. This is clearly related, since otherwise we force people to change `/properties/jenkins.version`, `/parent/version`, _and_ `/properties/java.level` in tandem. Anyway I suppose such a change should be very little work—just delete the property from `pom.xml` & ITs and hard-code `11` or `1.11` everywhere; optionally add a profile warning or failing if it is still defined. Anyone have more knowledge about this topic—whether there are plugins out there which _must_ run on 11+ (I guess due to some library); and what the UX is like if you try to install that plugin when Jenkins is running on 8? Perhaps we can at least switch the policy to default `java.level` to the currently supported version, ensure the build fails clearly if it is set to something too old, but allow it to be overridden to something newer in the rare case that is required?

One other question. If we use the incompatible approach here, do we release it once the first _weekly_ requiring Java 11 is released? Or wait for the first LTS after that? The former risks confusing devs with failing Dependabot PRs that would require a `jenkins.version` they are probably not interested in using yet; the latter might mean that plugins which do need a weekly dep would not have a working parent, unless they can use current parents with `<java.level>11</java.level>`. A middle way would be to release this as a `-beta` version which you can use if you know about it but which Dependabot will ignore.

File: pom.xml
Comment (jglick) [reply pom.xml:907]: Right, I misunderstood the intent of this PR.

**Comment by basil** [comment]
> I do recommend that as part of this

Is there any particular reason why cleaning up `java.level` needs to be done as part of this PR? Could not deleting `java.level` from `pom.xml` & ITs and hard-coding 8 or 1.8 everywhere be done in a separate PR and immediately merged/released without waiting for the Java 11 work?

**Comment by jglick** [comment]
> Is there any particular reason why cleaning up `java.level` needs to be done as part of this PR?

Indeed no; my advice was more that it be done no later than this PR, so that devs picking up a new baseline & this PR are not forced to explicitly edit POMs to select Java 11 when that is really their only option anyway.

> Could not deleting `java.level` […] be done in a separate PR and immediately merged/released without waiting for the Java 11 work?

Yes it could. Then this PR would just switch the hard-coded value from 8 to 11. Again pending some confirmation of real-world impact, that there is not a legitimate reason to override this to a different value than what would be implied by `jenkins.version`.

**Comment by jglick** [comment]
Fine with me. Anyway we have some months (I guess?) before this is even eligible for merge.

**Comment by basil** [comment]
> We do however need to ensure […] that the build fails and that the error message is comprehensible, so devs will be forced to look at the release notes and decide what to do.

This can be done in `org.jenkinsci.maven.plugins.hpi.ValidateMojo` by resolving `org.jenkins-ci.main:jenkins-bom:${JENKINS_VERSION}`, checking the `<properties>` section (currently the only encoding of the minimum Java version for core is in the `java.level` property, inherited from the parent POM, but we can always remove `java.level` from the parent POM add an explicit minimum Java version property to `jenkins-bom`), and then failing if the current JVM is too old.

> Probably this property should never have been defined.

Agreed.

> Perhaps we can deprecate it—ignore any value and issue a warning to that effect. Instead figure out what Java level to pass to tools based on the jenkins.version (somehow).

This _might_ be possible with a new `LifecyclePhase.INITIALIZE` Mojo that reads the minimum Java version as described above and then dynamically adjusts the Maven configuration prior to proceeding onto the next lifecycle phase. Possibly… :sweat_smile:

**Comment by jglick** [comment]
> a new `LifecyclePhase.INITIALIZE` Mojo that reads the minimum Java version as described above and then dynamically adjusts the Maven configuration

This would likely work for Maven builds as such, but could well confuse tools such as IDEs which expect to evaluate a POM to understand basic aspects of project configuration. (In particular, what `-source` or `-release` argument would be used for Java compilation, thus the syntax accepted and the APIs available.)

**Comment by basil** [comment]
Well I cannot come up with anything better so I am going to run with it.

**Comment by basil** [comment]
Yes, you read my mind regarding the last paragraph. That is exactly the conclusion I had come to.

> Detect the Java level based on `jenkins.version` as you seem to be proposing.

I thought it was you proposing that? :smile: I have lost track at this point. But on a high level I think there is a valid point here: the minimum Java version is conceptually tied to the Jenkins core version, so encoding it in the plugin POM (which is in principle agnostic to core version, within a supported range at least) seems like the wrong place. And as a result of this impedance mismatch, you end up with the problem of incompatible flag days and more.

Let me give another example of the above: it's not just a matter of Java level. With Java 17 support, we're adding `Add-Opens` directives to `MANIFEST.MF`, and ideally I'd like Surefire to run the tests with the same `Add-Opens` directives as the plugin's core version ships with. Sure, we could hard-code some Surefire configuration in the plugin POM, but that has the same problem as hard-coding a Java level: it creates flag days, since if the directives change in core (as they did recently) the plugin POM would need to be updated in lockstep, yet the plugin POM is supposed to be agnostic to the core version.

So I think dynamically setting things like Java level and Surefire arguments based on the core version makes sense at a high level. It avoids the flag days, and it avoids duplication of information. Core can encode official values of things like the minimum Java level and the `Add-Opens` directives _somewhere_ (e.g., in `core-bom`, or `MANIFEST.MF` in `jenkins.war`), and the build can then consume those based on the plugin's core version.

The challenge comes with tools, like you said. I wonder if we could blend both approaches: define a Java level and Surefire arguments in the plugin parent POM (corresponding to the minimum core version that the plugin parent POM supports) as a lowest common denominator and for the benefit of tools, but override those values dynamically in an initializer mojo based on the plugin's (not plugin POM's!) minimum core version. Maybe that way we could get the best of both worlds: precision for official builds, but degrading gracefully for GUI tools.

Speaking of which, do you think we have consensus on proceeding with https://github.com/jenkinsci/plugin-pom/pull/478#discussion_r809561384 (i.e., a revert of [JENKINS-20679](https://issues.jenkins.io/browse/JENKINS-20679))? I've prepared the commits for core, `plugin-pom`, `maven-hpi-plugin`, and `update-center2`, and they are extensive (hundreds of lines deleted), but I don't want to waste time testing and writing a description for Daniel about how backward compatibility will be maintained unless there is consensus that we want to go ahead with this.

**Comment by jglick** [comment]
> I thought it was you proposing that?

Yeah I did speculate idly in https://github.com/jenkinsci/plugin-pom/pull/478#pullrequestreview-886679297, though it does not seem like there is any way to do this in Maven Core (i.e. with a cleverly constructed POM). I am not even positive whether you can do this from a mojo early enough.

The “blend” approach would be nice if it worked. I am not too concerned about defining Surefire arguments dynamically since it is unlikely that tools would attempt to parse it out, unless there are some IDEs that bypass Maven to run tests yet read the POM (statically, without executing mojos) as the source of truth? But https://github.com/jenkinsci/plugin-pom/blob/d2b4d19d20a2db34e111e7181f5168f67696c704/pom.xml#L675-L678 seems like a bigger problem. Defaulting this to 11 would be fragile once we require 17 for core: plugins then depending on a version of core requiring 17 and actually using 17 APIs & language features would compile correctly from the shell (`mvn`) but would be scrambled in IDEs. I guess we would push out a POM update around the same time as the core release that switches the POM default, so IDEs would work again after a while.

IMO an incompatible release of the plugin POM is tolerable if it corresponds to a new Java major version requirement in core, which could arrive every couple years on average if we trail a fixed distance behind Java LTS releases. I agree that we would not want to be forcing a plugin POM range every time we adjust `Add-Opens`.

A more radical option would be to revert [JENKINS-32493](https://issues.jenkins.io/browse/JENKINS-32493): resume packaging the plugin pom as part of core. This would avoid the mismatch problems not just of `java.level` but things like https://github.com/jenkinsci/plugin-pom/blob/faf83a7bf03200b140fb111527e1879799e88e6e/pom.xml#L573-L585 but at the cost of slowing down updates to tools like `maven-hpi-plugin` and `jenkins-test-harness` and making it harder to test against alternate core versions in PCT and `buildPlugin()`.

> do you think we have consensus on […] a revert of JENKINS-20679 (plugins requesting a newer version of Java than the corresponding core)?

I am not confident at this point. I have not personally searched for examples of plugins doing this, nor do I know of any examples offhand.

**Comment by basil** [comment]
> though it does not seem like there is any way to do this in Maven Core (i.e. with a cleverly constructed POM).

That I am fairly certain about, having tried many ways of doing this. I do not think it is possible.

> I am not even positive whether you can do this from a mojo early enough.

Neither am I. Halfway through a prototype and it is challenging. :sweat_smile:

> The “blend” approach would be nice if it worked. […] I guess we would push out a POM update around the same time as the core release that switches the POM default, so IDEs would work again after a while. […] IMO an incompatible release of the plugin POM is tolerable if it corresponds to a new Java major version requirement in core, which could arrive every couple years on average if we trail a fixed distance behind Java LTS releases. I agree that we would not want to be forcing a plugin POM range every time we adjust `Add-Opens`.

Yeah agreed on all points.

> A more radical option would be to revert [JENKINS-32493](https://issues.jenkins.io/browse/JENKINS-32493): resume packaging the plugin pom as part of core.

Perhaps a bit too radical for me as part of this project. :smile:

> > do you think we have consensus on […] a revert of JENKINS-20679 (plugins requesting a newer version of Java than the corresponding core)?
> 
> I am not confident at this point. I have not personally searched for examples of plugins doing this, nor do I know of any examples offhand.

OK so what are the next steps for getting people to agree about changes like this? Developer list thread? It is really frustrating to attempt a large body of work only to get nitpicked to death in code review from people who don't entirely agree with every minute detail and have no constructive suggestions.

**Comment by jglick** [comment]
> what are the next steps for getting people to agree about changes like this?

I would say
1. Do basic research—at least grepping @jenkinsci plugins for apparent usages.
2. Post to the developer list proposing the change and explaining the impact.

In general the JEP process can be used to propose & document major changes with subtle impacts, but that seems like overkill for a system that possibly no one is really using to begin with.

**Comment by basil** [comment]
> Do basic research

I've done that already, Jesse. After all my work on Guava, do you really not trust me to do ecosystem wide code searches at this point? This comment is discouraging.

**Comment by jglick** [comment]
> I've done [research on JENKINS-20679] already

Oh, sorry, I was not aware. So there are no plugins on the UC which use a nonstandard Java dep? In that case it seems fine to me to delete the option to override `java.level`.

**Comment by basil** [comment]
> Oh, sorry, I was not aware. So there are no plugins on the UC which use a nonstandard Java dep?

So far I've only checked usages in sources but I was planning to check usages in binaries as part of "testing" the PRs. Actually, creating the commits was easy, but I anticipate the lion's share of the work will be testing (and writing PR descriptions in order to convince people to approve the PRs). So before sinking time in to do that testing and writing, I want to make sure people think this is a reasonable direction to go in. I will post to the developer list explaining my plan and, if there is notional agreement, do the additional writing and testing.

File: pom.xml
Comment (jglick) [inline pom.xml:677]: Nice, much simpler and clearer now.

**Comment by jglick** [comment]
95cd21ed7636a8f1ae122ebd8c30d288d5c829d9 seems to have reintroduced the `jdk-9-and-above` profile. Intentionally, or was this a confused merge and this profile can be reinlined back into the main POM configuration?

**Comment by basil** [comment]
> Intentionally […]?

This question implies that there is a problem with the PR (to be corrected by reinlining the profile). It does not state what the problem is or the reason why there is believed to be a problem. I will not participate in code reviews where reasoning is not provided.

**Comment by jglick** [comment]
> This question implies that there is a problem with the PR

I was asking whether the change was intentional. Especially if you are not using

```ini
[merge]
	conflictstyle = diff3
```

(which GitHub Web does not!) it is easy to apply a hunk to the wrong side. But maybe the change was done for a good reason and just got included in a merge commit without a separate commit message. If I follow correctly, until 95cd21ed7636a8f1ae122ebd8c30d288d5c829d9 (https://github.com/jenkinsci/plugin-pom/pull/478/files/22d4b09f2ea55dcf9b95e647a3ab2a401ccd0dee) we were switching (unconditionally, since no more `java.level`) from `-source 8 -target 8` to `-release 11` in the main POM config, and then also deleting the `jdk-above-9` profile under the assumption that the JDK used during build must be at least 11. Now the behavior on JDK 11+ would be mostly the same (there is also some `--add-opens`?), but this is moved into a profile so if you try to build using JDK 8 you would…get a build failure somehow? Since `maven-compiler-plugin` will default to 1.5 source level IIRC?

I am still confused about why this PR is closed. Is there some replacement elsewhere?

**Comment by jglick** [comment]
Merging against #530 & #531 I guess, so just reintroduced the JDK 9 profile but this _could_ be inlined for simplicity?

**Comment by basil** [comment]
> I was asking whether the change was intentional.

That question usually implies that the person asking the question believes there is a (potential) problem. If the person asking the question did not believe there was a (potential) problem, they would likely not be asking the question.

If you believe there is a problem with one of my PRs, please state what you believe the problem is and the reason why you believe it to be a problem. If the reason is valid, I will be happy to accommodate any changes. Do not ask me questions that imply there is a problem without stating it. Do not issue imperative statements requesting that I change X or Y without giving the reasoning for the request. I am being very clear about these boundaries.

> I am still confused about why this PR is closed. Is there some replacement elsewhere?

I could ask the same thing about https://github.com/jenkinsci/jenkins/pull/5304. As I wrote above, the reason this PR is closed is because you continue to comment on my PRs without providing reasoning. I am not interested in participating in such reviews.

File: pom.xml
Comment (jglick) [inline pom.xml:63]: openjdk/jdk11u-dev#919 FTR

**Pull request merged**
